From 66fbfb2aa6745b55fe8075df897ab9bbd279b13f Mon Sep 17 00:00:00 2001
From: Joseph Guo <qijian.guo@nxp.com>
Date: Fri, 13 Jun 2025 15:46:00 +0900
Subject: [PATCH 23/56] LF-15621: arm64: dts: imx93-11x11-frdm: add DT to
 support ov5640

Add imx93-11x11-frdm-ov5640.dts to support ov5640 camera for FRDM-IMX93.

Signed-off-by: Joseph Guo <qijian.guo@nxp.com>
Reviewed-by: Guoniu Zhou <guoniu.zhou@nxp.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |  3 +
 .../freescale/imx93-11x11-frdm-ov5640.dtso    | 70 +++++++++++++++++++
 2 files changed, 73 insertions(+)
 create mode 100644 arch/arm64/boot/dts/freescale/imx93-11x11-frdm-ov5640.dtso

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 9ddd9144cb76..28d37a25c6e8 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -497,6 +497,9 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk.dtb \
 dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-frdm.dtb imx93-11x11-frdm-8mic-reve.dtb \
 			  imx93-11x11-frdm-aud-hat.dtb imx93-11x11-frdm-ld.dtb
 
+imx93-11x11-frdm-ov5640-dtbs := imx93-11x11-frdm.dtb imx93-11x11-frdm-ov5640.dtbo
+dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-frdm-ov5640.dtb
+
 imx93-11x11-frdm-waveshare-7inch-c-panel-dtbs := imx93-11x11-frdm.dtb imx93-11x11-frdm-waveshare-7inch-c-panel.dtbo
 dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-frdm-waveshare-7inch-c-panel.dtb
 
diff --git a/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-ov5640.dtso b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-ov5640.dtso
new file mode 100644
index 000000000000..e3ebf166c02a
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-ov5640.dtso
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2025 NXP
+ */
+
+/dts-v1/;
+/plugin/;
+
+#include "imx93-pinfunc.h"
+#include <dt-bindings/gpio/gpio.h>
+
+&{/} {
+	clk_dummy: clock-dummy {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <27000000>;
+		clock-output-names = "clk_dummy";
+	};
+};
+
+&lpi2c3 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	/delete-node/ap1302_mipi@3c;
+
+	ov5640_mipi: ov5640_mipi@3c {
+		compatible = "ovti,ov5640";
+		reg = <0x3c>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_csi_pwdn>;
+		clocks = <&clk_dummy>;
+		clock-names = "xclk";
+		reset-gpios = <&pcal6524 22 GPIO_ACTIVE_LOW>;
+		powerdown-gpios = <&gpio4 28 GPIO_ACTIVE_HIGH>;
+		status = "okay";
+
+		port {
+			ov5640_mipi_ep: endpoint {
+				remote-endpoint = <&mipi_csi0_ep>;
+				data-lanes = <1 2>;
+				clock-lanes = <0>;
+			};
+		};
+	};
+};
+
+&mipi_csi {
+	status = "okay";
+
+	ports {
+		/delete-node/port@0;
+
+		port@0{
+			mipi_csi0_ep: endpoint {
+				remote-endpoint = <&ov5640_mipi_ep>;
+				cfg-clk-range = <28>;
+				hs-clk-range = <0x7>;
+				data-lanes = <1 2>;
+			};
+		};
+	};
+};
+
+&iomuxc {
+	pinctrl_csi_pwdn: csi_pwdn_grp{
+		fsl,pins = <
+			MX93_PAD_CCM_CLKO3__GPIO4_IO28	0x31e
+		>;
+	};
+};
-- 
2.34.1

